fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckPreferencesMiddleware.kt

```diff
@@ -0,0 +1,90 @@
+ */
+class ReviewQualityCheckPreferencesMiddleware(
+    private val reviewQualityCheckPreferences: ReviewQualityCheckPreferences,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+) : Middleware<ReviewQualityCheckState, ReviewQualityCheckAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<ReviewQualityCheckState, ReviewQualityCheckAction>,
+        next: (ReviewQualityCheckAction) -> Unit,
+        action: ReviewQualityCheckAction,
+    ) {
+        when (action) {
+            is ReviewQualityCheckAction.PreferencesMiddlewareAction -> processAction(action, next)
+
+            else -> {
+                // Forward other actions
+                next(action)
+            }
+        }
+    }
+
+    private fun processAction(
+        action: ReviewQualityCheckAction.PreferencesMiddlewareAction,
+        next: (ReviewQualityCheckAction) -> Unit,
+    ) {
+        when (action) {
+            is ReviewQualityCheckAction.Init -> {
+                scope.launch {
+                    val hasUserOptedIn = reviewQualityCheckPreferences.enabled()
+                    val isProductRecommendationsEnabled =
+                        reviewQualityCheckPreferences.productRecommendationsEnabled()
+                    next(
+                        ReviewQualityCheckAction.UpdateFeatureStateAction(
+                            hasUserOptedIn = hasUserOptedIn,
+                            isProductRecommendationsEnabled = isProductRecommendationsEnabled,
+                        ),
+                    )
+                }
+            }
+
+            is ReviewQualityCheckAction.OptInAction -> {
+                scope.launch {
+                    val isProductRecommendationsEnabled =
+                        reviewQualityCheckPreferences.productRecommendationsEnabled()
+                    next(
+                        ReviewQualityCheckAction.UpdateFeatureStateAction(
+                            hasUserOptedIn = action.hasUserOptedIn,
+                            isProductRecommendationsEnabled = isProductRecommendationsEnabled,
+                        ),
+                    )
+
+                    // Update the preference
+                    reviewQualityCheckPreferences.setEnabled(action.hasUserOptedIn)
+                }
+            }
+
+            is ReviewQualityCheckAction.ProductRecommendationAction -> {
+                next(
+                    ReviewQualityCheckAction.UpdateFeatureStateAction(
+                        hasUserOptedIn = true,
+                        isProductRecommendationsEnabled = action.isProductRecommendationEnabled,
+                    ),
+                )
+                scope.launch {
+                    // Update the preference
+                    reviewQualityCheckPreferences.setProductRecommendationsEnabled(
+                        action.isProductRecommendationEnabled,
+                    )
+                }
```

Comment:
should the coroutine usage instead live in `ReviewQualityCheckPreferencesImpl`? is it necessary?

Reply:
Updated to make the reading prefs, suspend functions, so it's more clear now, the scope is still injected in the store/middleware.